src/js/App/Sidenav/Navigation/DynamicNav.js

```diff
@@ -20,11 +20,25 @@ const DynamicNav = ({ dynamicNav }) => {
     if (navigation) {
       if (typeof navigation === 'function') {
         Promise.resolve(navigation({ schema, dynamicNav, currentNamespace })).then((data) => {
-          schema.navItems = schema?.navItems?.map((item, key) => ({
-            ...item,
-            ...(key === schema.navItems.findIndex((nav) => nav.dynamicNav === dynamicNav) && data),
-          }));
-          dispatch(loadLeftNavSegment(schema, currentNamespace, window.location.pathname));
+          const indexOfDynamicNav = schema.navItems.findIndex((nav) => nav.dynamicNav === dynamicNav);
+          delete schema.navItems[indexOfDynamicNav].dynamicNav;
+          if (Array.isArray(data)) {
+            schema.navItems.splice(
+              indexOfDynamicNav,
+              1,
+              ...data.map((item) => ({
+                appId: dynamicNav.split('/')[0],
+                ...schema.navItems[indexOfDynamicNav],
+                ...item,
+              }))
+            );
+          } else {
+            schema.navItems = schema?.navItems?.map((item, key) => ({
+              ...item,
+              ...(key === indexOfDynamicNav && data),
+            }));
+          }
+          dispatch(loadLeftNavSegment(schema, currentNamespace, window.location.pathname, true));
```

Comment:
`splice` and `schema.navItems =` mutate the schema (Redux) object. I think you want something that builds a new schema:

```
const start = schema.navItems.slice(0, Math.min(indexOfDynamicNav, 0));
const end = schema.navItems.slice(indexOfDynamicNav + 1);

const newContent = Array.isArray(data)
  ? data.map((item) => ({
      appId: dynamicNav.split('/')[0],
      ...schema.navItems[indexOfDynamicNav],
      ...item,
    }))
  : [data];

const updatedSchema = {
  ...schema,
  navItems: [
    ...start,
    ...newContent,
    ...end,
  ],
}

dispatch(loadLeftNavSegment(updatedSchema, currentNamespace, window.location.pathname, true));
```

Reply:
Yeah, good point. This shouldn't affect the schema, since we are mutating it right away. But it's better to do it this way.